addons/mail/static/src/rtc/voice_recorder.js

```diff
@@ -0,0 +1,124 @@
+import { useAttachmentUploader } from "@mail/attachments/attachment_uploader_hook";
+
+export class VoiceRecorder extends Component {
+    static props = ["composer"];
+    static template = "mail.VoiceRecorder";
+
+    microphone;
+    recorder;
+    startTimeStamp;
+
+    setup() {
+        this.state = useState({
+            isActionPending: false,
+            recording: false,
+            intervalId: false,
+            elapsed: "00",
+        });
+        this.notification = useService("notification");
+        this.userSettings = useService("mail.user_settings");
+        this.attachmentUploader = useAttachmentUploader(this.props.composer.thread, {
+            composer: this.props.composer,
+            isVoice: true,
+        });
+        this.isSafari = /^((?!chrome|android).)*safari/i.test(navigator.userAgent);
+        onWillUnmount(() => {
+            this.unmount = true;
+            this.stopRecording();
+        });
+    }
+
+    async startRecording() {
+        if (this.state.isActionPending) {
+            return;
+        }
+        this.state.isActionPending = true;
+        if (!this.microphone) {
+            try {
+                this.microphone = await browser.navigator.mediaDevices.getUserMedia({
+                    audio: this.userSettings.audioConstraints,
+                });
+            } catch {
+                this.notification.add(
+                    sprintf(_t('"%(hostname)s" needs to access your microphone'), {
+                        hostname: window.location.host,
+                    }),
+                    { type: "warning" }
+                );
+                this.state.isActionPending = false;
+                return;
+            }
+        }
+        this.state.recording = true;
+        let totalSeconds = 0;
+        this.state.intervalId = setInterval(() => {
+            if (totalSeconds == 60) {
+                this.stopRecording();
+                this.notification.add(_t("The Maximum recording duration could be one minute"), {
+                    type: "warning",
+                });
+            }
+            ++totalSeconds;
+            this.state.elapsed = totalSeconds < 10 ? "0" + totalSeconds : totalSeconds;
+        }, 1000);
+        const config = {
+            mimeType: this.isSafari ? "audio/mp4" : "audio/webm",
+        };
+        this.recorder = new MediaRecorder(this.microphone, config);
+        this.startTimeStamp = Date.now();
+        this.recorder.ondataavailable = (e) => {
+            if (this.unmount) {
+                return;
+            }
+            const blob = new Blob([e.data], {
+                type: this.recorder.mimeType || config.mimeType,
+            });
+            const duration = e.timecode - this.startTimeStamp;
+            const fileExtension = config.mimeType === "audio/mp4" ? "mp4" : "webm";
+            const file = new File([blob], this.filename(fileExtension), {
+                type: config.mimeType,
+            });
```

Comment:
I think the file extension could be directly calculated inside the filename function (based directly on `this.isSafari`), the param then become useless.